File: js/test/utils/sineStar.ts
/**
 * SineStar polygon generator and precision reducer.
 * Ported from JTS SineStarFactory.java and GeometryPrecisionReducer.
 */
import type { Polygon, Position } from "geojson";

/**
 * Generate a SineStar polygon (port of JTS SineStarFactory.createSineStar).
 *
 * Creates a polygon shaped like a star with sinusoidal arms.
 * The polygon has `nPts` vertices distributed around a circle,
 * with the radius modulated by a sine wave to create arm-like protrusions.
 */
export function createSineStar(
  centreX: number,
  centreY: number,
  size: number,
  nPts: number,
  nArms: number,
  armRatio: number,
): Polygon {
  const radius = size / 2.0;
  const clampedRatio = Math.max(0.0, Math.min(1.0, armRatio));
  const armMaxLen = clampedRatio * radius;
  const insideRadius = (1 - clampedRatio) * radius;

  const coords: Position[] = [];
  for (let i = 0; i < nPts; i++) {
    const ptArcFrac = (i / nPts) * nArms;
    const armAngFrac = ptArcFrac - Math.floor(ptArcFrac);
    const armAng = 2 * Math.PI * armAngFrac;
    const armLenFrac = (Math.cos(armAng) + 1.0) / 2.0;
    const curveRadius = insideRadius + armMaxLen * armLenFrac;

    const ang = i * ((2 * Math.PI) / nPts);
    const x = curveRadius * Math.cos(ang) + centreX;
    const y = curveRadius * Math.sin(ang) + centreY;
    coords.push([x, y]);
  }
  // Close the ring
  coords.push([coords[0][0], coords[0][1]]);

  return { type: "Polygon", coordinates: [coords] };
}

/**
 * Reduce coordinate precision (port of JTS GeometryPrecisionReducer).
 * Rounds each coordinate to: round(coord * scale) / scale
 */
export function reducePrecision(polygon: Polygon, scale: number): Polygon {
  const reduced = polygon.coordinates.map((ring) =>
    ring.map(([x, y]) => [Math.round(x * scale) / scale, Math.round(y * scale) / scale] as Position),
  );
  return { type: "Polygon", coordinates: reduced };
}

File: rs/benches/utils/sine_star.rs
//! SineStar polygon generator and precision reducer.
//! Ported from JTS SineStarFactory.java and GeometryPrecisionReducer.

use geo_types::{Coord, LineString, Polygon};
use std::f64::consts::PI;

/// Generate a SineStar polygon (port of JTS SineStarFactory.createSineStar).
///
/// Creates a polygon shaped like a star with sinusoidal arms.
pub fn create_sine_star(
    centre_x: f64,
    centre_y: f64,
    size: f64,
    n_pts: usize,
    n_arms: usize,
    arm_ratio: f64,
) -> Polygon<f64> {
    let radius = size / 2.0;
    let clamped_ratio = arm_ratio.clamp(0.0, 1.0);
    let arm_max_len = clamped_ratio * radius;
    let inside_radius = (1.0 - clamped_ratio) * radius;

    let mut coords = Vec::with_capacity(n_pts + 1);
    for i in 0..n_pts {
        let pt_arc_frac = (i as f64 / n_pts as f64) * n_arms as f64;
        let arm_ang_frac = pt_arc_frac - pt_arc_frac.floor();
        let arm_ang = 2.0 * PI * arm_ang_frac;
        let arm_len_frac = (arm_ang.cos() + 1.0) / 2.0;
        let curve_radius = inside_radius + arm_max_len * arm_len_frac;

        let ang = i as f64 * (2.0 * PI / n_pts as f64);
        let x = curve_radius * ang.cos() + centre_x;
        let y = curve_radius * ang.sin() + centre_y;
        coords.push(Coord { x, y });
    }
    // Close the ring
    coords.push(coords[0]);

    Polygon::new(LineString(coords), vec![])
}

/// Reduce coordinate precision (port of JTS GeometryPrecisionReducer).
/// Rounds each coordinate to: round(coord * scale) / scale
pub fn reduce_precision(polygon: &Polygon<f64>, scale: f64) -> Polygon<f64> {
    let reduce = |c: &Coord<f64>| Coord {
        x: (c.x * scale).round() / scale,
        y: (c.y * scale).round() / scale,
    };

    let exterior: Vec<Coord<f64>> = polygon.exterior().0.iter().map(reduce).collect();
    let interiors: Vec<LineString<f64>> = polygon
        .interiors()
        .iter()
        .map(|ring| LineString(ring.0.iter().map(reduce).collect()))
        .collect();

    Polygon::new(LineString(exterior), interiors)
}
